README.md
# tio - a simple TTY terminal I/O application

[![Build Status](https://travis-ci.org/tio/tio.svg?branch=master)](https://travis-ci.org/tio/tio)
[![Snap Status](https://build.snapcraft.io/badge/tio/tio.snapcraft.svg)](https://build.snapcraft.io/user/tio/tio.snapcraft)

## 1. Introduction

tio is a simple TTY terminal application which features a straightforward
commandline interface to easily connect to TTY devices for basic input/output.

It was created because the author needed a simple no-nonsense TTY terminal
application to easily connect to various terminal TTY devices.

<p align="center">
<img src="https://tio.github.io/images/tio-demo.gif">
</p>


## 2. Usage

The commandline interface is straightforward as reflected in the output from
'tio --help':
```
    Usage: tio [<options>] <tty-device>

    Options:
      -b, --baudrate <bps>        Baud rate (default: 115200)
      -d, --databits 5|6|7|8      Data bits (default: 8)
      -f, --flow hard|soft|none   Flow control (default: none)
      -s, --stopbits 1|2          Stop bits (default: 1)
      -p, --parity odd|even|none  Parity (default: none)
      -o, --output-delay <ms>     Output delay (default: 0)
      -n, --no-autoconnect        Disable automatic connect
      -e, --local-echo            Do local echo
      -l, --log <filename>        Log to file
      -m, --map <flags>           Map special characters
      -v, --version               Display version
      -h, --help                  Display help

    See the man page for list of supported mapping flags.

    In session, press ctrl-a q to quit.
```

The only option which requires a bit of elaboration is perhaps the
`--no-autoconnect` option.

By default tio automatically connects to the provided device if present.  If
the device is not present, it will wait for it to appear and then connect. If
the connection is lost (eg. device is unplugged), it will wait for the device
to reappear and then reconnect. However, if the `--no-autoconnect` option is
provided, tio will exit if the device is not present or an established
connection is lost.

Tio features full bash autocompletion support.

Tio also supports various key commands. Press ctrl-a ? to list the available
key commands.

See the tio man page for more details.


## 3. Installation

The latest release version is available at https://tio.github.io

### 3.1 Installation using release tarball

Install steps:
```
     $ ./configure
     $ make
     $ make install
```
See INSTALL file for more installation details.

### 3.2 Installation using package

Tio comes prepackaged for various GNU/Linux distributions. Visit
https://tio.github.io for package installation details.


## 4. Contributing

Tio is open source. All contributions (bug fixes, doc, ideas, etc.) are
welcome. Visit the tio GitHub page to access latest source code, create pull
requests, add issues etc..

GitHub: https://github.com/tio/tio

Also, if you find this free open source software useful please consider making
a donation:

[![Donate](https://www.paypal.com/en_US/i/btn/x-click-but21.gif)](https://www.paypal.me/lundmar)


## 5. Support

Submit bug reports via GitHub: https://github.com/tio/tio/issues


## 6. Website

Visit https://tio.github.io


## 7. License

Tio is GPLv2+. See COPYING file for license details.


## 8. Authors

Created by Martin Lund \<martin.lund@keep-it-simple.com>

See the AUTHORS file for full list of authors.


ChangeLog

=== tio v1.32 ===



Changes since tio v1.31:

 * Update AUTHORS

 * Minor code style cleanups

 * Cleanup print macros

 * Flush output

   Make sure output is transmitted immediately by flushing the output.

Robey Pointer:

 * add optional timestamps

   with "-t" or "C-t T", toggle a timestamp prefix to each line.

Jakub Wilk:

 * Fix typos

Sylvain Lafrasse:

 * Added macOS compatibility

 * Made O_NONBLOCK flag to open() call specific to macOS only.

 * Added macOS-related details.

 * Added O_NONBLOCK flag to open() call for macOS (10.13.6) compatibility.



Changes since tio v1.30:

 * Update date

 * Update AUTHORS

Henner Zeller:

 * Clarify the input/output variable names (No-op change)

 * Organize options the same sequence they are mentioned in cmdline help.

 * Update README.

 * Map CR->NL locally on output instead of using tio.c_oflag |= OCRNL.

   This mostly is intended to have local echo output exactly what is sent
   to the remote endpoint.
   A nice side-effect is, that it also fixes tty-implementations, that can't
   deal with the OCRNL flag on tio.c_oflag.

 * Provide local-echo option.

   Can be switched on with -e on the command line.
   Can be toggled with Ctrl t e while program is running.

 * Write to logfile as soon as we have the data, don't buffer.

   Logfiles are important to see what happened, in particular if something
   unexpected happened; so we want to make sure that the logfile is flushed
   to disk.

   Before this change, the logfile was typically written at the end in
   a large chunk as the default (large) buffering applied. Now, characters are
   written out ASAP, so it is possible to get a live-view with a
   tail -f <logfile>



Changes since tio v1.29:

 * Update README

 * Update man page and bash completion

 * Update AUTHORS

qianfan Zhao:

 * ONLCRNL: change the method to map NL to CR-NL



Changes since tio v1.28:

 * Add mapping flags INLCRNL and ODELBS

   The following new mapping flags are added:

   INLCRNL: Map NL to CR-NL on input.
   ODELBS: Map DEL to BS on output.

   Flags requested and tested by Jan Ciger (janoc).



Changes since tio v1.27:

 * Update README

 * Update AUTHORS

 * Add snap status to README.md

 * Add README.md to prettify GitHub page

 * Add missing header

Petr Vaněk:

 * Add missing header file under musl-libc

   Musl's inclusion tree slightly differs from glibc, therefore TCGETS2 is
   not reachable through sys/ioctl.h, so asm/ioctls.h needs to be included
   too.

Jakub Wilk:

 * Fix grammar and typos



Changes since tio v1.26:

 * Update man page

 * Add support for setting non-standard baudrates

   Support for non-standard baudrate settings will be automatically enabled
   if the termios2 interface is detected available. However, to play it
   safe, the old and widely supported termios interface will still be used
   when setting standard baudrates.

 * Cleanup

 * Update AUTHORS



Changes since tio v1.25:

 * Reconfigure stdin

   Make stdin behave more raw'ish. In particular, don't
   translate CR -> NL on input.

 * Add special character map feature

   Add a --map option which allows mapping special characters, in particular CR and
   NL characters which are used in various combinations on various platforms.

 * Cleanup

 * Update AUTHORS

 * Update README

 * Mention website

 * Update man page



Changes since tio v1.24:

 * Fix error applying new stdout settings

   On Fedora 26 tio will quit with the following error message:

   "Error: Could not apply new stdout settings (Invalid argument)"

   In case of Fedora, it turns out that the new stdout settings used are a
   bit too aggressive because an empty termios structure is used. To remedy
   this we reuse the existing stdout settings and only reconfigure the
   specific options we need to make a "raw" stdout configuration.

 * Remove unused pkgconfig in configure

 * Code cleanup

   Remove unused variable.



Changes since tio v1.23:

 * Optimize clear screen command

   Replaced system call with inline ANSI/VT100 clear screen code sequence

 * Fix bash completion installation

   Fixed the configure script to avoid that the bash completion script gets
   installed outside of the prefix location. The default install location
   is now $prefix/share/bash-completion/completions.

   Use the configure option '--with-bash-completion-dir=PATH' if you need
   to install the bash completion script elsewhere.

Jakub Wilk:

 * Add missing commas in conditional sentences



Changes since tio v1.22:

 * Update copyright headers

Jakub Wilk:

 * Fix typos



Changes since tio v1.21:

 * Update man page date

 * Update copyright year

 * Code cleanup

 * Update README and man page



Changes since tio v1.20:

 * Add support for hexadecimal mode

   A new key command 'ctrl-a h' is introduced which toggles between
   hexadecimal mode and normal mode. When in hexadecimal mode data received
   will be printed in hexadecimal.

 * Do not distribute src/bash_completion/tio

   Since the bash completion tio script is now autogenerated from tio.in it
   should not be distributed in the tarball.

 * Add missing forward flag

 * Update AUTHORS file

Adam Borowski:

 * 'ctrl-a b' to send serial break.

Jakub Wilk:

 * Removed git commit references from ChangeLog

   ChangeLog is primary useful for users who don't have the git repository
   at hand.

   Replace git commit references with version numbers; or if the change
   only cleans up another change with no release in between, remove the
   changelog item completely.



Changes since tio v1.19:

 * Added more error handling of terminal calls

   Also removed duplicate terminal flushing calls.

 * Revert "Added support for non-standard baud rates"

   This reverts a change made in v1.18.

   Reverting because supporting non-standard or arbitrary baud rates is
   troublesome because the c library provides no means of doing so and even
   if bare metal linux kernel interface is used it will not work on all

src/include/tio/tty.h
/*
 * tio - a simple TTY terminal I/O application
 *
 * Copyright (c) 2014-2017  Martin Lund
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef TTY_H
#define TTY_H

#define KEY_QUESTION 0x3f
#define KEY_B 0x62
#define KEY_C 0x63
#define KEY_E 0x65
#define KEY_H 0x68
#define KEY_L 0x6C
#define KEY_Q 0x71
#define KEY_S 0x73
#define KEY_T 0x74
#define KEY_SHIFT_T 0x54
//#define KEY_CTRL_T 0x14
#define KEY_CTRL_T 0x1
#define KEY_V 0x76

#define NORMAL 0
#define HEX 1

void stdout_configure(void);
void stdout_restore(void);
void stdin_configure(void);
void stdin_restore(void);
void tty_configure(void);
int tty_connect(void);
void tty_wait_for_device(void);

#endif


src/options.c
/*
 * tio - a simple TTY terminal I/O application
 *
 * Copyright (c) 2014-2018  Martin Lund
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "config.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <errno.h>
#include <getopt.h>
#include <termios.h>
#include <limits.h>
#include "tio/options.h"
#include "tio/error.h"

/* Default options */
struct option_t option =
{
    "",       // Device name
    115200,   // Baudrate
    8,        // Databits
    "none",   // Flow
    1,        // Stopbits
    "none",   // Parity
    0,        // No output delay
    false,    // No autoconnect
    false,    // No log
    false,    // No local echo
    false,    // No timestamp
    "",       // Log filename
    ""        // Map string
};

void print_help(char *argv[])
{
    printf("Usage: %s [<options>] <tty-device>\n", argv[0]);
    printf("\n");
    printf("Options:\n");
    printf("  -b, --baudrate <bps>        Baud rate (default: 115200)\n");
    printf("  -d, --databits 5|6|7|8      Data bits (default: 8)\n");
    printf("  -f, --flow hard|soft|none   Flow control (default: none)\n");
    printf("  -s, --stopbits 1|2          Stop bits (default: 1)\n");
    printf("  -p, --parity odd|even|none  Parity (default: none)\n");
    printf("  -o, --output-delay <ms>     Output delay (default: 0)\n");
    printf("  -n, --no-autoconnect        Disable automatic connect\n");
    printf("  -e, --local-echo            Do local echo\n");
    printf("  -t, --timestamp             Prefix each new line with a timestamp\n");
    printf("  -l, --log <filename>        Log to file\n");
    printf("  -m, --map <flags>           Map special characters\n");
    printf("  -v, --version               Display version\n");
    printf("  -h, --help                  Display help\n");
    printf("\n");
    printf("See the man page for list of supported mapping flags.\n");
    printf("\n");
    printf("In session, press ctrl-a q to quit.\n");
    printf("\n");
}

long string_to_long(char *string)
{
    long result;
    char *end_token;

    errno = 0;
    result = strtol(string, &end_token, 10);
    if ((errno != 0) || (*end_token != 0))
    {
        error_printf("Invalid digit");
        exit(EXIT_FAILURE);
    }

    return result;
}

void parse_options(int argc, char *argv[])
{
    int c;

    if (argc == 1)
    {
        print_help(argv);
        exit(EXIT_SUCCESS);
    }

    while (1)
    {
        static struct option long_options[] =
        {
            {"baudrate",       required_argument, 0, 'b'},
            {"databits",       required_argument, 0, 'd'},
            {"flow",           required_argument, 0, 'f'},
            {"stopbits",       required_argument, 0, 's'},
            {"parity",         required_argument, 0, 'p'},
            {"output-delay",   required_argument, 0, 'o'},
            {"no-autoconnect", no_argument,       0, 'n'},
            {"local-echo",     no_argument,       0, 'e'},
            {"timestamp",      no_argument,       0, 't'},
            {"log",            required_argument, 0, 'l'},
            {"map",            required_argument, 0, 'm'},
            {"version",        no_argument,       0, 'v'},
            {"help",           no_argument,       0, 'h'},
            {0,                0,                 0,  0 }
        };

        /* getopt_long stores the option index here */
        int option_index = 0;

        /* Parse argument using getopt_long */
        c = getopt_long(argc, argv, "b:d:f:s:p:o:netl:m:vh", long_options, &option_index);

        /* Detect the end of the options */
        if (c == -1)
            break;

        switch (c)
        {
            case 0:
                /* If this option sets a flag, do nothing else now */
                if (long_options[option_index].flag != 0)
                    break;
                printf("option %s", long_options[option_index].name);
                if (optarg)
                    printf(" with arg %s", optarg);
                printf("\n");
                break;

            case 'b':
                option.baudrate = string_to_long(optarg);
                break;

            case 'd':
                option.databits = string_to_long(optarg);
                break;

            case 'f':
                option.flow = optarg;
                break;

            case 's':
                option.stopbits = string_to_long(optarg);
                break;

            case 'p':
                option.parity = optarg;
                break;

            case 'o':
                option.output_delay = string_to_long(optarg);
                break;

            case 'n':
                option.no_autoconnect = true;
                break;

            case 'e':
                option.local_echo = true;
                break;

            case 't':
                option.timestamp = true;
                break;

            case 'l':
                option.log = true;
                option.log_filename = optarg;
                break;

            case 'm':
                option.map = optarg;
                break;

            case 'v':
                printf("tio v%s\n", VERSION);
                printf("Copyright (c) 2014-2018 Martin Lund\n");
                printf("\n");
                printf("License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl-2.0.html>.\n");
                printf("This is free software: you are free to change and redistribute it.\n");
                printf("There is NO WARRANTY, to the extent permitted by law.\n");
                exit(EXIT_SUCCESS);
                break;

            case 'h':
                print_help(argv);
                exit(EXIT_SUCCESS);
                break;

            case '?':
                /* getopt_long already printed an error message */
                exit(EXIT_FAILURE);
                break;

            default:
                exit(EXIT_FAILURE);
        }
    }

    /* Assume first non-option is the tty device name */
    if (optind < argc)
        option.tty_device = argv[optind++];

    if (strlen(option.tty_device) == 0)
    {
        error_printf("Missing device name");
        exit(EXIT_FAILURE);
    }

    /* Print any remaining command line arguments (unknown options) */
    if (optind < argc)
    {
        printf("Error: unknown arguments: ");
        while (optind < argc)
            printf("%s ", argv[optind++]);
        printf("\n");
        exit(EXIT_FAILURE);
    }
}
